File: 1531-string-compression-ii/1531-string-compression-ii.cpp
class Solution {
private:
    unordered_map<int, int> memo;
    set<int> add{1, 9, 99};
    int dp(string& s, int i, char lastChar, int lastCharCount, int k){
        if (k < 0)
            return INT_MAX / 2;
        
        if (i == s.size())
            return 0;
        
        int key = i * 101 * 27 * 101 + (lastChar - 'a') * 101 * 101 + lastCharCount * 101 + k;
        if (memo.find(key) != memo.end())
            return memo[key];
        
        int keepChar;
        int deleteChar = dp(s, i + 1, lastChar, lastCharCount, k - 1);
        if (s[i] == lastChar)
            keepChar = dp(s, i + 1, lastChar, lastCharCount + 1, k) + add.count(lastCharCount);
        else
            keepChar = dp(s, i + 1, s[i], 1, k) + 1;
        int ans = min(keepChar, deleteChar);
        return memo[key] = ans;
    }
    
public:
    int getLengthOfOptimalCompression(string s, int k) {
        return dp(s, 0, (char)('a' + 26), 0, k);
    }
};

File: 1531-string-compression-ii/1531-string-compression-ii.py
class Solution:
    def getLengthOfOptimalCompression(self, s: str, k: int) -> int:
        def digits(x: int) -> int:
            return 1 if x == 1 else 2 if 1 < x < 10 else 3 if 10 <= x < 100 else 4
        
        n = len(s)
        dp = [[float('inf')] * (k + 1) for _ in range(n + 1)]
        dp[0][0] = 0
        
        for i in range(1, n + 1):
            for j in range(min(i, k) + 1):
                if j > 0:
                    dp[i][j] = dp[i - 1][j - 1]
                
                remove = count = 0
                for left in range(i, 0, -1):
                    if s[left - 1] == s[i - 1]:
                        count += 1
                    else:
                        remove += 1
                        if remove > j:
                            break
                    
                    dp[i][j] = min(dp[i][j], dp[left - 1][j - remove] + digits(count))
        
        return dp[-1][-1]
                